odl/solvers/nonsmooth/difference_convex.py

```diff
@@ -0,0 +1,243 @@
+            callback(x)
+
+
+def doubleprox_dc(x, y, g, h, phi, K, niter, gamma, mu, callback=None):
+    r"""Double-proxmial gradient d.c. algorithm of Banert and Bot.
+
+    This algorithm solves a problem of the form::
+
+        min_x g(x) + phi(x) - h(Kx).
+
+    Parameters
+    ----------
+    x : `LinearSpaceElement`
+        Initial primal guess, updated in-place.
+    y : `LinearSpaceElement`
+        Initial dual guess, updated in-place.
+    g : `Functional`
+        The nonsmooth part of the convex part of the problem. Needs to provide
+        a `Functional.proximal` factory.
+    h : `Functional`
+        The functional involved in the concave part of the problem. Needs to
+        provide a `Functional.proximal` factory.
+    phi : `Functional`
+        The smooth part of the convex part of the problem. Needs to provide a
+        `Functional.gradient`, and convergence can be guaranteed if the
+        gradient is Lipschitz continuous.
+    K : `Operator`
+        The operator involved in the concave part of the problem.
+    niter : int
+        Number of iterations.
+    gamma : positive float
+        Stepsize in the primal updates.
+    mu : positive float
+        Stepsize in the dual updates.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Notes
+    -----
+    This algorithm is proposed in`[BB2016]
+    <https://arxiv.org/abs/1610.06538>`_ and solves the d.c. problem
+
+    .. math ::
+        \min_x g(x) + \varphi(x) - h(Kx)
+
+    together with its Toland dual
+
+    .. math ::
+        \min_y h^*(y) - (g + \varphi)^*(K^* y).
+
+    The iterations are given by
+
+    .. math ::
+        x_{n+1} &= \mathrm{Prox}_{\gamma}^g (x_n + \gamma K^* y_n
+                   - \gamma \nabla \varphi(x_n)), \\
+        y_{n+1} &= \mathrm{Prox}_{\mu}^{h^*} (y_n + \mu K x_{n+1}).
+
+    To guarantee convergence, the parameter :math:`\gamma` must satisfy
+    :math:`0 < \gamma < 2/L` where :math:`L` is the Lipschitz constant of
+    :math:`\nabla \varphi`.
+
+    References
+    ----------
+    [BB2016] Banert, S, and Bot, R I. *A general double-proximal gradient
+    algorithm for d.c. programming*. arXiv:1610.06538 [math.OC] (2016).
+    """
+    primal_space = g.domain
+    dual_space = h.domain
+
+    if phi.domain != primal_space:
+        raise ValueError('`g.domain` and `phi.domain` need to be equal, but '
+                         '{} != {}'.format(primal_space, phi.domain))
+    if K.domain != primal_space:
+        raise ValueError('`g.domain` and `K.domain` need to be equal, but '
+                         '{} != {}'.format(primal_space, K.domain))
+    if K.range != dual_space:
+        raise ValueError('`h.domain` and `K.range` need to be equal, but '
+                         '{} != {}'.format(dual_space, K.range))
+
+    for _ in range(niter):
+        g.proximal(gamma)(x + gamma * K.adjoint(y) -
+                          gamma * phi.gradient(x), out=x)
```

Comment:
Without temporaries, you can make this faster by using `gamma * (K.adjoint(y) - phi.gradient(x))`. You can also use `x.lincomb(1, x, gamma, K.adjoint(y) - phi.gradient(x))` for even more efficiency. Since it returns `x` you can use it as expression in the argument of the prox.